nanomq/sub_handler.c

```diff
@@ -146,12 +148,22 @@ encode_suback_msg(nng_msg *msg, nano_work *work)
 
 #if SUPPORT_MQTT5_0
 	if (PROTOCOL_VERSION_v5 == proto_ver) { // add property in variable
-		//TODO set property if necessary 
-		// encode_properties(msg, sub_pkt->properties);
 		encode_properties(msg, NULL, CMD_SUBACK);
 	}
 #endif
 
+	// Note. packetid should be non-zero, BUT in order to make subclients
+	// known that, we return an error(ALREADY IN USE)
+	reason_code = PACKET_IDENTIFIER_IN_USE;
+	if (sub_pkt->packet_id == 0) {
+		if ((rv = nng_msg_append(msg, &reason_code, 1)) != 0) {
+			debug_msg("ERROR: nng_msg_append [%d]", rv);
+			return PROTOCOL_ERROR;
+		}
+	}
+
+	// Note. When packet_id is zero, node must be empty. So, Dont worry
+	// the order of reasone codes would be changed.
```

Comment:
reason

Reply:
That's how I design it. Nodes would not be allocated if packet id is equal to zero. So we just respond an ack to remind the client that the packet id is wrong.